File: iOS/Pulsar/Sources/Pulsar/AudioSimulator.swift
import Foundation
import AVFoundation

public enum WaveformType: String {
	case sine = "sine"
	case square = "square"
	case triangle = "triangle"
	case sawtooth = "sawtooth"
}

struct OscillatorConfig {
	let frequency: (initial: Double, final: Double, decayTime: Double)
	let envelope: (attack: Double, decay: Double, sustainLevel: Double, sustainDuration: Double, release: Double)
	let waveform: WaveformType
}

struct DiscreteAudioConfig {
	let oscillator: OscillatorConfig
	let timestamp: Double
	let volume: Float
}

struct ContinuousAudioConfig {
	let type: String
	let data: (amplitude: [ChartPoint], frequency: [ChartPoint])
}

struct AudioPatternConfig {
	let discreteData: [DiscreteAudioConfig]
	let continuousData: [ContinuousAudioConfig]
}

public class AudioSimulator: NSObject {
	private let sampleRate: Double = 44100
	private var audioContext: AVAudioEngine = AVAudioEngine()
	private var offlineContext: AVAudioEngine?
	private var renderedBuffer: AVAudioPCMBuffer?
	private var currentSource: AVAudioPlayerNode?
	private var playerNode: AVAudioPlayerNode = AVAudioPlayerNode()
	private var filterNode: AVAudioUnitEQ = AVAudioUnitEQ(numberOfBands: 1)
	private var isInitialized = false
	private var isEngineConfigured = false
	private var currentConfig: AudioPatternConfig?
    
	public override init() {
		super.init()
		configureAudioContext()
	}

	public func parsePattern(from data: PlaygroundData) {
		renderedBuffer = nil
		
		currentConfig = AudioPatternConfig(
      discreteData: generateDiscreteAudioConfig(from: data),
      continuousData: generateContinuousAudioConfig(from: data)
    )
		
		renderPattern(config: currentConfig!)
	}
    
	private func configureAudioContext() {
		guard !isEngineConfigured else { return }
		
    let session = AVAudioSession.sharedInstance()
		do {
			try session.setCategory(.playback, mode: .default, options: [.mixWithOthers])
			try session.setActive(true)
		} catch {
			print("AudioSession error: \(error)")
		}
        
		// Setup audio engine: Player -> Filter(lowpass) -> MainMixer
		audioContext.attach(playerNode)
        
		let band = filterNode.bands.first!
		band.filterType = .lowPass
		band.frequency = 700
		band.bandwidth = 1.0
		band.gain = 1
		band.bypass = false
		audioContext.attach(filterNode)
        
		let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!
		audioContext.connect(playerNode, to: filterNode, format: format)
		audioContext.connect(filterNode, to: audioContext.mainMixerNode, format: format)
        
		do {
			try audioContext.start()
			isEngineConfigured = true
		} catch {
			print("Failed to start AVAudioEngine: \(error)")
		}
	}
	
	private func generateWaveform(_ waveform: WaveformType, phase: Double) -> Double {
		let normalizedPhase = phase.truncatingRemainder(dividingBy: Double.pi * 2) / (Double.pi * 2)
		
		switch waveform {
		case .sine:
			return sin(phase)
		case .square:
			return normalizedPhase < 0.5 ? 1.0 : -1.0
		case .triangle:
			if normalizedPhase < 0.25 {
				return normalizedPhase * 4.0
			} else if normalizedPhase < 0.75 {
				return 2.0 - (normalizedPhase * 4.0)
			} else {
				return (normalizedPhase - 1.0) * 4.0
			}
		case .sawtooth:
			return (normalizedPhase * 2.0) - 1.0
		}
	}
    
	private func generateDiscreteAudioConfig(from data: PlaygroundData) -> [DiscreteAudioConfig] {
		var discreteData: [DiscreteAudioConfig] = []
		
		let sources = 3
		let maxFrequency = 440.0
		let minFrequency = 60.0
		
		func normalizeFrequency(_ value: Float) -> Double {
			return minFrequency + (maxFrequency - minFrequency) * Double(value)
		}
		
		func alignVolume(_ x: Float, sources: Int) -> Float {
			return Float(0.1 / Double(sources) + 0.9 / Double(sources) * Double(x))
		}
		
		for bar in data.bar {
			let baseFrequency = normalizeFrequency(bar.y2)
			let targetFrequency = baseFrequency * 0.2
			let volume = alignVolume(bar.y1, sources: sources)
			
			// Base oscillator
			discreteData.append(DiscreteAudioConfig(
				oscillator: OscillatorConfig(
					frequency: (initial: baseFrequency, final: targetFrequency, decayTime: 0.028),
					envelope: (attack: 0.002, decay: 0, sustainLevel: 1, sustainDuration: 0, release: 0.014),
					waveform: .sine
				),
				timestamp: Double(bar.x) * 1000,
				volume: volume
			))
			
			// Harmonic 1 (1.5x)
			let harmonic1 = baseFrequency * 1.5
			let targetHarmonic1 = harmonic1 * 0.4
			discreteData.append(DiscreteAudioConfig(
				oscillator: OscillatorConfig(
					frequency: (initial: harmonic1, final: targetHarmonic1, decayTime: 0.031),
					envelope: (attack: 0, decay: 0, sustainLevel: 1, sustainDuration: 0, release: 0.015),
					waveform: .sine
				),
				timestamp: Double(bar.x) * 1000,
				volume: volume
			))
			
			// Harmonic 2 (0.3x)
			let harmonic2 = baseFrequency * 0.3
			let targetHarmonic2 = harmonic2 * 0.5
			discreteData.append(DiscreteAudioConfig(
				oscillator: OscillatorConfig(
					frequency: (initial: harmonic2, final: targetHarmonic2, decayTime: 0.039),
					envelope: (attack: 0.005, decay: 0, sustainLevel: 1, sustainDuration: 0, release: 0.018),
					waveform: .sine
				),
				timestamp: Double(bar.x) * 1000,
				volume: volume
			))
		}
		
		return discreteData
	}
	
	private func generateContinuousAudioConfig(from data: PlaygroundData) -> [ContinuousAudioConfig] {
		let amplitudePoints: [ChartPoint] = data.line.count > 0 ? data.line[0] : []
		let frequencyPoints: [ChartPoint] = data.line.count > 1 ? data.line[1] : []
		
		func normalizeFrequency(_ x: Float) -> Double {
			return 80.0 + (230.0 - 80.0) * Double(x)
		}
		
		func applyModifiers(amplitude: [ChartPoint], frequency: [ChartPoint], ampMod: Double, freqMod: Double) -> (amplitude: [ChartPoint], frequency: [ChartPoint]) {
			let modifiedAmplitude = amplitude.map { point in
				ChartPoint(x: point.x, y: Float(Double(point.y) * ampMod))
			}
			let modifiedFrequency = frequency.map { point in
				ChartPoint(x: point.x, y: Float(normalizeFrequency(point.y) * freqMod))
			}
			return (modifiedAmplitude, modifiedFrequency)
		}
		
		var continuousConfigs: [ContinuousAudioConfig] = [
      ContinuousAudioConfig(
        type: "sine",
        data: applyModifiers(amplitude: amplitudePoints, frequency: frequencyPoints, ampMod: 0.6, freqMod: 0.8),
      ),
      ContinuousAudioConfig(
        type: "triangle",
        data: applyModifiers(amplitude: amplitudePoints, frequency: frequencyPoints, ampMod: 0.2, freqMod: 0.4),
      ),
      ContinuousAudioConfig(
        type: "sine",
        data: applyModifiers(amplitude: amplitudePoints, frequency: frequencyPoints, ampMod: 0.5, freqMod: 1.0),
      )
    ]
		
		return continuousConfigs
	}
	
	private func renderPattern(config: AudioPatternConfig) {
		let discreteData = config.discreteData
		let continuousData = config.continuousData
    
    func calculateTotalDuration() -> Double {
      var continuousDuration: Double = 0
      for wave in continuousData {
        if !wave.data.amplitude.isEmpty {
          continuousDuration = max(continuousDuration, wave.data.amplitude.last!.x)
        }
      }
      continuousDuration += 0.01
      
      var discreteDuration: Double = 0
      for config in discreteData {
        let eventStartTime = config.timestamp / 1000.0
        let envelope = config.oscillator.envelope
        let oscillatorDuration = envelope.attack + envelope.decay + envelope.sustainDuration + envelope.release
        let eventEndTime = eventStartTime + oscillatorDuration
        discreteDuration = max(discreteDuration, eventEndTime)
      }
      discreteDuration += 0.1
      
      return max(continuousDuration, discreteDuration)
    }
		
    let totalDuration = calculateTotalDuration()
		let frameCount = AVAudioFrameCount(Int(totalDuration * sampleRate) + 1)
		let format = AVAudioFormat(standardFormatWithSampleRate: sampleRate, channels: 1)!
		guard let buffer = AVAudioPCMBuffer(pcmFormat: format, frameCapacity: frameCount) else {
			renderedBuffer = nil
			return
		}
		buffer.frameLength = frameCount
		let out = buffer.floatChannelData![0]
        
		// Initialize buffer to silence
		memset(out, 0, Int(frameCount) * MemoryLayout<Float>.size)
        
		var phasesContinuous: [Double] = Array(repeating: 0, count: continuousData.count)
		var phasesDiscrete: [Double] = Array(repeating: 0, count: discreteData.count)
		let twoPi = Double.pi * 2
        
		func valueForTime(_ points: [ChartPoint], _ t: Double) -> Float {
			if points.isEmpty { return 0 }
			if t <= points[0].x { return points[0].y }
			if t >= points.last!.x { return points.last!.y }
			for i in 1..<points.count {
				let p1 = points[i - 1]
				let p2 = points[i]
				if t <= p2.x {
					let t0 = p1.x, t1 = p2.x
					let v0 = p1.y, v1 = p2.y
					let ratio = Float((t - t0) / (t1 - t0))
					return v0 + (v1 - v0) * ratio
				}
			}
			return points.last!.y
		}
        
		func normalizeFrequency(_ y: Float) -> Double {
			return Double(y)
		}
		
		for i in 0..<Int(frameCount) {
			let t = Double(i) / sampleRate
			var sampleValue: Double = 0
            
			// Continuous waves
			for (waveIdx, waveConfig) in continuousData.enumerated() {
				if !waveConfig.data.amplitude.isEmpty && !waveConfig.data.frequency.isEmpty {
					let amp = Double(valueForTime(waveConfig.data.amplitude, t))
					let freq = normalizeFrequency(valueForTime(waveConfig.data.frequency, t))
                    
					phasesContinuous[waveIdx] += twoPi * freq / sampleRate
					if phasesContinuous[waveIdx] > twoPi { phasesContinuous[waveIdx] -= twoPi }
					
					let waveformType = WaveformType(rawValue: waveConfig.type) ?? .sine
					sampleValue += amp * generateWaveform(waveformType, phase: phasesContinuous[waveIdx])
				}
			}
            
			// Discrete waves (transients)
			for (oscIdx, oscConfig) in discreteData.enumerated() {
				let eventStartTime = oscConfig.timestamp / 1000.0
				let relativeTime = t - eventStartTime
				
				guard relativeTime >= 0 else { continue }
				
				let envelope = oscConfig.oscillator.envelope
				let totalDuration = envelope.attack + envelope.decay + envelope.sustainDuration + envelope.release
				
				guard relativeTime < totalDuration else { continue }
				guard oscIdx < phasesDiscrete.count else { continue }
                
				// Apply envelope (ADSR)
				var envValue: Float = 0
				if relativeTime < envelope.attack {
					envValue = envelope.attack > 0 ? Float(relativeTime / envelope.attack) : 1.0
				} else if relativeTime < envelope.attack + envelope.decay {
					envValue = 1.0
				} else if relativeTime < envelope.attack + envelope.decay + envelope.sustainDuration {
					envValue = Float(envelope.sustainLevel)
				} else {
					let releaseTime = relativeTime - (envelope.attack + envelope.decay + envelope.sustainDuration)
					envValue = envelope.release > 0 ? Float(1.0 - (releaseTime / envelope.release)) : 0
				}
                
				// Frequency sweep
				let freqConfig = oscConfig.oscillator.frequency
				var freq = freqConfig.initial
				if freqConfig.decayTime > 0 {
					let sweepDuration = min(freqConfig.decayTime, totalDuration)
					if relativeTime < sweepDuration {
						let ratio = relativeTime / freqConfig.decayTime
						freq = freqConfig.initial * pow(freqConfig.final / freqConfig.initial, ratio)
					} else {
						freq = freqConfig.final
					}
				}
                
				phasesDiscrete[oscIdx] += twoPi * freq / sampleRate
				if phasesDiscrete[oscIdx] > twoPi { phasesDiscrete[oscIdx] -= twoPi }
                
				sampleValue += Double(oscConfig.volume * envValue) * generateWaveform(oscConfig.oscillator.waveform, phase: phasesDiscrete[oscIdx])
			}
            
			out[i] = Float(sampleValue)
		}
        
		renderedBuffer = buffer
	}
  
	public func play() {
		guard let buffer = renderedBuffer else { return }
		configureAudioContext()
        
		if playerNode.isPlaying { playerNode.stop() }
		playerNode.scheduleBuffer(buffer, at: nil, options: []) { [weak self] in
			DispatchQueue.main.async {
				self?.stop()
			}
		}
		playerNode.play()
	}
    
	public func stop() {
		playerNode.stop()
	}
    
	public var isPlaying: Bool {
		return playerNode.isPlaying
	}
}

File: react-native/react-native-pulsar/android/src/main/java/com/swmansion/pulsar/audio/AudioSimulator.kt
package com.swmansion.pulsar.audio

import android.media.AudioAttributes
import android.media.AudioFormat
import android.media.AudioTrack
import com.swmansion.pulsar.BuildConfig
import com.swmansion.pulsar.types.AudioDataConfig
import com.swmansion.pulsar.types.AudioPatternConfig
import com.swmansion.pulsar.types.CompatibilityMode
import com.swmansion.pulsar.types.ContinuousAudioConfig
import com.swmansion.pulsar.types.DiscreteAudioConfig
import com.swmansion.pulsar.types.EnvelopeConfig
import com.swmansion.pulsar.types.FrequencyConfig
import com.swmansion.pulsar.types.OscillatorConfig
import com.swmansion.pulsar.types.PatternData
import com.swmansion.pulsar.types.ValuePoint
import com.swmansion.pulsar.types.WaveformType
import kotlin.math.*
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock

class AudioSimulator(
    private var compatibilityMode: CompatibilityMode
) {
    companion object {
        private const val MAX_FREQUENCY = 440.0
        private const val MIN_FREQUENCY = 60.0
        private const val CONTINUOUS_FREQUENCY_MIN = 80.0
        private const val CONTINUOUS_FREQUENCY_MAX = 230.0
        private const val NUM_SOURCES = 3
    }

    private val sampleRate: Int = 22050
    private val sampleRateF: Float = sampleRate.toFloat()
    private var audioTrack: AudioTrack? = null
    private var isInitialized = false
    private var playSound: Boolean = BuildConfig.DEBUG
    private val audioScope = CoroutineScope(Dispatchers.IO)
    private val audioMutex = Mutex()

    fun parsePattern(data: PatternData): ByteArray? {
        if (!playSound) return null

        val currentConfig = AudioPatternConfig(
            discreteData = generateDiscreteAudioConfig(data),
            continuousData = generateContinuousAudioConfig(data)
        )

        return renderPattern(currentConfig)
    }

    fun enableSound(value: Boolean) {
        this.playSound = value
    }

    fun play(buffer: ByteArray?) {
        if (buffer == null || !playSound || buffer.isEmpty()) return

        audioScope.launch {
            audioMutex.withLock {
                try {
                    if (audioTrack == null) {
                        configureAudioContext()
                    }

                    audioTrack?.apply {
                        if (playState == AudioTrack.PLAYSTATE_PLAYING) {
                            stop()
                            flush()
                        }

                        play()

                        var totalBytesWritten = 0
                        var lastWriteTime = System.currentTimeMillis()

                        while (totalBytesWritten < buffer.size) {
                            val bytesToWrite = buffer.size - totalBytesWritten
                            val bytesWritten = write(buffer, totalBytesWritten, bytesToWrite)

                            if (bytesWritten < 0) {
                                break
                            } else if (bytesWritten == 0) {
                                val currentTime = System.currentTimeMillis()
                                if (currentTime - lastWriteTime > 5000) {
                                    break
                                }
                                delay(10)
                            } else {
                                totalBytesWritten += bytesWritten
                                lastWriteTime = System.currentTimeMillis()
                            }
                        }
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    fun stop() {
        if (!playSound) {
            return
        }
        audioScope.launch {
            audioMutex.withLock {
                try {
                    audioTrack?.apply {
                        if (playState == AudioTrack.PLAYSTATE_PLAYING) {
                            stop()
                            flush()
                        }
                    }
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        }
    }

    private fun configureAudioContext() {
        if (isInitialized && audioTrack != null) return

        val minBufferSize = AudioTrack.getMinBufferSize(
            sampleRate,
            AudioFormat.CHANNEL_OUT_MONO,
            AudioFormat.ENCODING_PCM_16BIT
        )

        audioTrack = AudioTrack.Builder()
            .setAudioAttributes(
                AudioAttributes.Builder()
                    .setUsage(AudioAttributes.USAGE_MEDIA)
                    .setContentType(AudioAttributes.CONTENT_TYPE_SONIFICATION)
                    .build()
            )
            .setAudioFormat(
                AudioFormat.Builder()
                    .setEncoding(AudioFormat.ENCODING_PCM_16BIT)
                    .setSampleRate(sampleRate)
                    .setChannelMask(AudioFormat.CHANNEL_OUT_MONO)
                    .build()
            )
            .setBufferSizeInBytes(minBufferSize * 4)
            .setTransferMode(AudioTrack.MODE_STREAM)
            .build()

        isInitialized = true
    }

    private fun generateWaveform(waveform: WaveformType, phase: Float): Float {
        val normalizedPhase = (phase % (2f * PI.toFloat())) / (2f * PI.toFloat())

        return when (waveform) {
            WaveformType.SINE -> sin(phase)
            WaveformType.SQUARE -> if (normalizedPhase < 0.5f) 1.0f else -1.0f
            WaveformType.TRIANGLE -> when {
                normalizedPhase < 0.25f -> normalizedPhase * 4.0f
                normalizedPhase < 0.75f -> 2.0f - (normalizedPhase * 4.0f)
                else -> (normalizedPhase - 1.0f) * 4.0f
            }
            WaveformType.SAWTOOTH -> (normalizedPhase * 2.0f) - 1.0f
        }
    }

    private fun normalizeFrequency(value: Float): Double {
        return MIN_FREQUENCY + (MAX_FREQUENCY - MIN_FREQUENCY) * value
    }

    private fun alignVolume(x: Float, sources: Int): Float {
        return (0.1 / sources + 0.9 / sources * x).toFloat()
    }

    private fun generateDiscreteAudioConfig(data: PatternData): List<DiscreteAudioConfig> {
        val configs = mutableListOf<DiscreteAudioConfig>()

        data.discretePattern.forEach { discretePoint ->
            val baseFrequency = normalizeFrequency(discretePoint.frequency)
            val targetFrequency = baseFrequency * 0.2
            val volume = alignVolume(discretePoint.amplitude, NUM_SOURCES)

            // Base oscillator
            configs.add(
                DiscreteAudioConfig(
                    oscillator = OscillatorConfig(
                        frequency = FrequencyConfig(
                            initial = baseFrequency,
                            final = targetFrequency,
                            decayTime = 0.028
                        ),
                        envelope = EnvelopeConfig(
                            attack = 0.002,
                            decay = 0.0,
                            sustainLevel = 1.0,
                            sustainDuration = 0.0,
                            release = 0.014
                        ),
                        waveform = WaveformType.SINE
                    ),
                    timestamp = discretePoint.time.toDouble(),
                    volume = volume
                )
            )

            // Harmonic 1 (1.5x)
            val harmonic1 = baseFrequency * 1.5
            val targetHarmonic1 = harmonic1 * 0.4
            configs.add(
                DiscreteAudioConfig(
                    oscillator = OscillatorConfig(
                        frequency = FrequencyConfig(
                            initial = harmonic1,
                            final = targetHarmonic1,
                            decayTime = 0.031
                        ),
                        envelope = EnvelopeConfig(
                            attack = 0.0,
                            decay = 0.0,
                            sustainLevel = 1.0,
                            sustainDuration = 0.0,
                            release = 0.015
                        ),
                        waveform = WaveformType.SINE
                    ),
                    timestamp = discretePoint.time.toDouble(),
                    volume = volume
                )
            )

            // Harmonic 2 (0.3x)
            val harmonic2 = baseFrequency * 0.3
            val targetHarmonic2 = harmonic2 * 0.5
            configs.add(
                DiscreteAudioConfig(
                    oscillator = OscillatorConfig(
                        frequency = FrequencyConfig(
                            initial = harmonic2,
                            final = targetHarmonic2,
                            decayTime = 0.039
                        ),
                        envelope = EnvelopeConfig(
                            attack = 0.005,
                            decay = 0.0,
                            sustainLevel = 1.0,
                            sustainDuration = 0.0,
                            release = 0.018
                        ),
                        waveform = WaveformType.SINE
                    ),
                    timestamp = discretePoint.time.toDouble(),
                    volume = volume
                )
            )
        }

        return configs
    }

    private fun generateContinuousAudioConfig(data: PatternData): List<ContinuousAudioConfig> {
        val amplitudePoints = data.continuousPattern.amplitude.ifEmpty { emptyList() }
        val frequencyPoints = data.continuousPattern.frequency.ifEmpty { emptyList() }

        fun normalizeFrequencyForContinuous(x: Float): Float {
            return (CONTINUOUS_FREQUENCY_MIN + (CONTINUOUS_FREQUENCY_MAX - CONTINUOUS_FREQUENCY_MIN) * x).toFloat()
        }

        fun applyModifiers(
            amplitude: List<ValuePoint>,
            frequency: List<ValuePoint>,
            ampMod: Float,
            freqMod: Float
        ): Pair<List<ValuePoint>, List<ValuePoint>> {
            val modifiedAmplitude = amplitude.map { point ->
                ValuePoint(point.time, point.value * ampMod)
            }
            val modifiedFrequency = frequency.map { point ->
                ValuePoint(point.time, (normalizeFrequencyForContinuous(point.value) * freqMod))
            }
            return Pair(modifiedAmplitude, modifiedFrequency)
        }

        val configs = mutableListOf<ContinuousAudioConfig>()

        val (amp1, freq1) = applyModifiers(amplitudePoints, frequencyPoints, 0.6f, 0.8f)
        configs.add(ContinuousAudioConfig(type = "sine", data = AudioDataConfig(amplitude = amp1, frequency = freq1)))

        val (amp2, freq2) = applyModifiers(amplitudePoints, frequencyPoints, 0.2f, 0.4f)
        configs.add(ContinuousAudioConfig(type = "triangle", data = AudioDataConfig(amplitude = amp2, frequency = freq2)))

        val (amp3, freq3) = applyModifiers(amplitudePoints, frequencyPoints, 0.5f, 1.0f)
        configs.add(ContinuousAudioConfig(type = "sine", data = AudioDataConfig(amplitude = amp3, frequency = freq3)))

        return configs
    }

    private fun renderPattern(config: AudioPatternConfig): ByteArray? {
        val totalDuration = calculateTotalDuration(config)
        val bufferSize = (totalDuration * sampleRate).toInt() + 1
        if (bufferSize <= 0) return null

        val samples = FloatArray(bufferSize)
        val phasesContinuous = FloatArray(config.continuousData.size)
        val phasesDiscrete = FloatArray(config.discreteData.size)
        val twoPi = 2f * PI.toFloat()

        val continuousWaveforms: Array<WaveformType> = config.continuousData.map { wave ->
            WaveformType.entries.find { it.name.equals(wave.type, ignoreCase = true) } ?: WaveformType.SINE
        }.toTypedArray()

        data class DiscreteOscillatorCache(
            val startSample: Int,
            val endSample: Int,
            val freqInitial: Float,
            val freqFinal: Float,
            val freqDecaySamples: Int,
            val logFreqRatio: Float,
            val envAttackSamples: Int,
            val envDecayEndSamples: Int,
            val envSustainEndSamples: Int,
            val envTotalSamples: Int,
            val sustainLevel: Float,
            val volume: Float,
            val waveform: WaveformType
        )

        val discreteCache: List<DiscreteOscillatorCache> = config.discreteData.map { osc ->
            val env = osc.oscillator.envelope
            val totalEnvDur = env.attack + env.decay + env.sustainDuration + env.release
            val startSample = (osc.timestamp / 1000.0 * sampleRate).toInt()
            val endSample = startSample + ceil(totalEnvDur * sampleRate).toInt()
            val freqCfg = osc.oscillator.frequency
            val sweepDur = if (freqCfg.decayTime > 0) minOf(freqCfg.decayTime, totalEnvDur) else 0.0
            val logRatio = if (freqCfg.decayTime > 0 && freqCfg.initial > 0)
                ln(freqCfg.final / freqCfg.initial).toFloat() else 0f
            DiscreteOscillatorCache(
                startSample = startSample,
                endSample = endSample,
                freqInitial = freqCfg.initial.toFloat(),
                freqFinal = freqCfg.final.toFloat(),
                freqDecaySamples = (sweepDur * sampleRate).toInt(),
                logFreqRatio = logRatio,
                envAttackSamples = (env.attack * sampleRate).toInt(),
                envDecayEndSamples = ((env.attack + env.decay) * sampleRate).toInt(),
                envSustainEndSamples = ((env.attack + env.decay + env.sustainDuration) * sampleRate).toInt(),
                envTotalSamples = (totalEnvDur * sampleRate).toInt(),
                sustainLevel = env.sustainLevel.toFloat(),
                volume = osc.volume,
                waveform = osc.oscillator.waveform
            )
        }

        val continuousAmpCursors = IntArray(config.continuousData.size) { 1 }
        val continuousFreqCursors = IntArray(config.continuousData.size) { 1 }

        fun valueForTime(points: List<ValuePoint>, tMs: Float, cursors: IntArray, idx: Int): Float {
            if (points.isEmpty()) return 0f
            if (tMs <= points[0].time) return points[0].value
            if (tMs >= points.last().time) return points.last().value
            while (cursors[idx] < points.size && points[cursors[idx]].time < tMs) cursors[idx]++
            val p1 = points[cursors[idx] - 1]
            val p2 = points[cursors[idx]]
            val denom = p2.time - p1.time
            if (denom <= 0f) return p1.value
            val ratio = (tMs - p1.time) / denom
            return p1.value + (p2.value - p1.value) * ratio
        }

        for (i in 0 until bufferSize) {
            val tMs = i.toFloat() / sampleRateF * 1000f
            var sampleValue = 0f

            // Continuous waves
            config.continuousData.forEachIndexed { waveIdx, waveConfig ->
                if (waveConfig.data.amplitude.isNotEmpty() && waveConfig.data.frequency.isNotEmpty()) {
                    val amp = valueForTime(waveConfig.data.amplitude, tMs, continuousAmpCursors, waveIdx)
                    val freq = valueForTime(waveConfig.data.frequency, tMs, continuousFreqCursors, waveIdx)
                        .coerceIn(20f, sampleRateF / 2f)

                    phasesContinuous[waveIdx] += twoPi * freq / sampleRateF
                    if (phasesContinuous[waveIdx] > twoPi) phasesContinuous[waveIdx] -= twoPi

                    sampleValue += amp * generateWaveform(continuousWaveforms[waveIdx], phasesContinuous[waveIdx])
                }
            }

            discreteCache.forEachIndexed { oscIdx, pre ->
                if (i < pre.startSample || i >= pre.endSample) return@forEachIndexed

                val relSample = i - pre.startSample

                val envValue: Float = when {
                    relSample < pre.envAttackSamples ->
                        if (pre.envAttackSamples > 0) relSample.toFloat() / pre.envAttackSamples else 1f
                    relSample < pre.envDecayEndSamples -> 1f
                    relSample < pre.envSustainEndSamples -> pre.sustainLevel
                    else -> {
                        val relRelease = relSample - pre.envSustainEndSamples
                        val releaseSamples = pre.envTotalSamples - pre.envSustainEndSamples
                        if (releaseSamples > 0) 1f - relRelease.toFloat() / releaseSamples else 0f
                    }
                }

                val freq = if (pre.freqDecaySamples > 0 && relSample < pre.freqDecaySamples) {
                    val ratio = relSample.toFloat() / pre.freqDecaySamples
                    pre.freqInitial * exp(pre.logFreqRatio * ratio)
                } else {
                    pre.freqFinal
                }

                phasesDiscrete[oscIdx] += twoPi * freq / sampleRateF
                if (phasesDiscrete[oscIdx] > twoPi) phasesDiscrete[oscIdx] -= twoPi

                sampleValue += pre.volume * envValue * generateWaveform(pre.waveform, phasesDiscrete[oscIdx])
            }

            samples[i] = sampleValue
        }

        return convertToByteArray(samples)
    }

    private fun calculateTotalDuration(config: AudioPatternConfig): Double {
        var continuousDuration = 0.0
        for (wave in config.continuousData) {
            if (wave.data.amplitude.isNotEmpty()) {
                continuousDuration = maxOf(continuousDuration, wave.data.amplitude.last().time.toDouble() / 1000.0)
            }
        }
        continuousDuration += 0.01

        var discreteDuration = 0.0
        for (cfg in config.discreteData) {
            val eventStartTime = cfg.timestamp / 1000.0
            val envelope = cfg.oscillator.envelope
            val oscillatorDuration = envelope.attack + envelope.decay + envelope.sustainDuration + envelope.release
            val eventEndTime = eventStartTime + oscillatorDuration
            discreteDuration = maxOf(discreteDuration, eventEndTime)
        }
        discreteDuration += 0.1

        return maxOf(continuousDuration, discreteDuration)
    }

    private fun convertToByteArray(samples: FloatArray): ByteArray {
        val byteArray = ByteArray(samples.size * 2)
        for (i in samples.indices) {
            val sample = (samples[i] * 32767f).toInt().coerceIn(-32768, 32767)
            byteArray[i * 2] = (sample and 0xFF).toByte()
            byteArray[i * 2 + 1] = ((sample shr 8) and 0xFF).toByte()
        }
        return byteArray
    }

    fun setCompatibilityMode(mode: CompatibilityMode) {
        compatibilityMode = mode
    }
}
